src/FSharp.Core/reflect.fs

```diff
@@ -679,69 +658,137 @@ module internal Impl =
-        | 3 -> table.[2].MakeGenericType tys
-        | 4 -> table.[3].MakeGenericType tys
-        | 5 -> table.[4].MakeGenericType tys
-        | 6 -> table.[5].MakeGenericType tys
-        | 7 -> table.[6].MakeGenericType tys
-        | n when n >= maxTuple ->
-            let tysA = tys.[0 .. tupleEncField - 1]
-            let tysB = tys.[maxTuple - 1 ..]
-            let tyB = mkTupleType isStruct asm tysB
-            table.[7].MakeGenericType(Array.append tysA [| tyB |])
-        | _ -> invalidArg "tys" (SR.GetString(SR.invalidTupleTypes))
+    module internal ObsoleteTupleTypeMaking =
+        let private tupleNames =
+            [|
+                "System.Tuple`1"
+                "System.Tuple`2"
+                "System.Tuple`3"
+                "System.Tuple`4"
+                "System.Tuple`5"
+                "System.Tuple`6"
+                "System.Tuple`7"
+                "System.Tuple`8"
+                "System.Tuple"
+                "System.ValueTuple`1"
+                "System.ValueTuple`2"
+                "System.ValueTuple`3"
+                "System.ValueTuple`4"
+                "System.ValueTuple`5"
+                "System.ValueTuple`6"
+                "System.ValueTuple`7"
+                "System.ValueTuple`8"
+                "System.ValueTuple"
+            |]
+
+        let private dictionaryLock = obj ()
+        let private refTupleTypes = Dictionary<Assembly, Type[]>()
+        let private valueTupleTypes = Dictionary<Assembly, Type[]>()
+
+        let rec mkTupleType isStruct (asm: Assembly) (tys: Type[]) =
+            let table =
+                let makeIt n =
+                    let tupleFullName n =
+                        let structOffset = if isStruct then 9 else 0
+                        let index = n - 1 + structOffset
+                        tupleNames.[index]
+
+                    match n with
+                    | 1 -> asm.GetType(tupleFullName 1)
+                    | 2 -> asm.GetType(tupleFullName 2)
+                    | 3 -> asm.GetType(tupleFullName 3)
+                    | 4 -> asm.GetType(tupleFullName 4)
+                    | 5 -> asm.GetType(tupleFullName 5)
+                    | 6 -> asm.GetType(tupleFullName 6)
+                    | 7 -> asm.GetType(tupleFullName 7)
+                    | 8 -> asm.GetType(tupleFullName 8)
+                    | _ -> invalidArg "tys" (SR.GetString(SR.invalidTupleTypes))
+
+                let tables =
+                    if isStruct then
+                        valueTupleTypes
+                    else
+                        refTupleTypes
+
+                match lock dictionaryLock (fun () -> tables.TryGetValue asm) with
+                | false, _ ->
+                    // the Dictionary<>s here could be ConcurrentDictionary<>'s, but then
+                    // that would lock while initializing the Type array (maybe not an issue)
+                    let mutable a = Array.init<Type> 8 (fun i -> makeIt (i + 1))
```

Comment:
concurrentdictionary is the way to go here.

Reply:
This is the old code, just moved. I did not touch it a single bit.